### PR TITLE
Add a "minimal crate metadata" API endpoint

### PR DESCRIPTION
**Pull request by teozkr**
As mentioned in #4503 and https://github.com/renovatebot/renovate/issues/3486, this is intended to be a lighter-weight API than `/crates/:crate_id`, which can be subject to looser rate limits.

This adds a new `/crates/:crate_id/crate` endpoint for metadata about the crate itself, without information about related objects (such as versions or categories).

### CODE REVIEW

**Comment by teozkr** [comment]
7256d79 and 817f87a are tangential housekeeping tasks, feel free to merge without them (or I can split them into a separate PR, if you prefer that).

**Comment by lfrancke** [comment]
Thanks @teozkr 

Yes, the rate limit is not enforced in code.

Let's see what others say, this adds an additional endpoint while the original compromise called for an additional parameter on the existing `show` endpoint.

**Review by Turbo87 — COMMENTED** [review]
> I can split them into a separate PR, if you prefer that

that would be great, yeah 👍  ideally two separate PRs since the two changes are itself unrelated too :)

> This adds a new `/crates/:crate_id/crate` endpoint for metadata about the crate itself, without information about related objects (such as versions or categories).

IMHO having two endpoints for the same resource makes it a bit confusing. was there a reason why you went that way instead of optimizing the existing endpoint?

could you also make sure to add a few tests for the new code? 🙏

**Comment by teozkr** [comment]
> IMHO having two endpoints for the same resource makes it a bit confusing. was there a reason why you went that way instead of optimizing the existing endpoint?

To me, "overview of X and subresources" and "info about X" are two fairly separate code paths. A client would also want to have completely separate type definitions for them, for example.

There's also the server code structure issue, that doing it as a separate flag for `show` would end up cluttering `show` with a lot of branching, while also making it less clear what queries are actually performed for each mode (important both for validating this PR in particular, and to ensure that the basic mode doesn't regress with future features).

**Comment by teozkr** [comment]
Added a test based on the full metadata endpoint test. Ideally it'd validate that we only make one DB query, but it doesn't look like Diesel has a way to query for that, as far as I can tell.

**Comment by Turbo87** [comment]
> To me, "overview of X and subresources" and "info about X" are two fairly separate code paths.

I tend to disagree. That endpoint is returning a resource and in the default case it currently includes a bunch of related resources (aka. relationships). The suggestion in the in linked issue was to make that automatic relationship inclusion opt-out. Having two different endpoint for the same resource may make it easier right now to implement things, but it also means we have to maintain two different endpoints that basically do the same thing. I'd prefer to be able to use something like https://jsonapi.org/format/#fetching-includes, even though we're currently not using the JSON:API schema for our responses.

File: src/controllers/krate/metadata.rs [hunk 1]
```diff
@@ -111,71 +112,183 @@ pub fn summary(req: &mut dyn RequestExt) -> EndpointResult {
...
+            categories: true,
+            badges: true,
+            downloads: true,
+        }
+    }
+}
+
+impl ShowIncludeMode {
+    const INVALID_COMPONENT: &'static str =
+        "invalid component for ?mode= (expected 'versions', 'keywords', 'categories', 'badges', 'downloads', or 'full')";
```
[the remark this answers is not in the document]
Comment (teozkr) [reply src/controllers/krate/metadata.rs]: Oof, must have missed that one when renaming it. Should be fixed now.

**Comment by soenkeliebau** [comment]
This looks excellent, thanks!

**Comment by jtgeibel** [comment]
Thanks!

@bors r+

**Pull request merged**